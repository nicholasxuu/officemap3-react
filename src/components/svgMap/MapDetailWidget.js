import React from 'react';
import { PropTypes } from 'react';
import Immutable from 'immutable';
import * as ImmutablePropTypes from 'react-immutable-proptypes';
import '../../styles/svgMap/mapDetailWidget.css';
import { Popover } from 'react-bootstrap';

class MapDetailWidget extends React.Component {

	render = () => {
		if (this.props.show === false) {
			return null;
		}

		return (
			<Popover
				id="map-detail-widget"
				className="map-detail-widget"
				placement="top"
				positionLeft={this.props.pagePosX}
				positionTop={this.props.pagePosY}
				title={this.props.location.get('name')}
			    style={{
			    }}
			>
				<div className="widget-container">
					<div className="widget-image-container">
						<img
							className="widget-image"
							src={this.props.location.get('image')}
						    alt={this.props.location.get('name')}
						/>
					</div>
					<div className="widget-detail">
						{this.props.location.get('description')}
					</div>
				</div>
			</Popover>
		);
	}
}

MapDetailWidget.defaultProps = {
	show: false,
	pagePosX: 0,
	pagePosY: 0,
	location: Immutable.Map({
		id: 0,
		name: 'dummy',
	}),
};

MapDetailWidget.propTypes = {
	show: PropTypes.bool.isRequired,
	pagePosX: PropTypes.number.isRequired,
	pagePosY: PropTypes.number.isRequired,
	location: ImmutablePropTypes.contains({
		id: PropTypes.number,
		name: PropTypes.string,
	}).isRequired,
};

export default MapDetailWidget;

