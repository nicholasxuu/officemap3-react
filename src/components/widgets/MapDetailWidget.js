import React from 'react';
import { PropTypes } from 'react';
import Immutable from 'immutable';
import * as ImmutablePropTypes from 'react-immutable-proptypes';
import '../../styles/widgets/mapDetailWidget.css';
import { Popover } from 'react-bootstrap';

class MapDetailWidget extends React.Component {

	renderImage = () => {
		if (!this.props.location.get('image')) {
			return null;
		}

		return (
			<div className="widget-image-container">
				<img
					className="widget-image"
					src={this.props.location.get('image')}
					alt={this.props.location.get('name')}
					style={{
						display: 'block',
						overflow: 'hidden',
						width: '100px',
						maxHeight: '100%',
						objectFit: 'cover',
					}}
				/>
			</div>
		);
	};

	render = () => {
		if (this.props.show === false) {
			return null;
		}

		const imageDom = this.renderImage();

		return (
			<Popover
				id="map-detail-widget"
				className="map-detail-widget"
				placement="top"
				positionLeft={this.props.pagePosX}
				positionTop={this.props.pagePosY}
				title={this.props.location.get('name')}
			    style={{
				    zIndex: 3,
				    /* make widget top and center */
				    transform: 'translateX(-50%) translateY(-100%)',
			    }}
			>
				<div
					className="widget-container"
				    style={{
					    display: 'flex',
					    flexFlow: 'row nowrap',
					    overflow: 'hidden',
					    width: '100%',
					    height: '100%',
				    }}
				>
					{imageDom}
					<div
						className="widget-detail"
					    style={{
						    display: 'block',
						    overflow: 'hidden',
						    width: '200px',
					    }}
					>
						{this.props.location.get('description')}
					</div>
				</div>
			</Popover>
		);
	}
}

MapDetailWidget.defaultProps = {
	show: false,
	pagePosX: 0,
	pagePosY: 0,
	location: Immutable.Map({
		id: 0,
		name: 'dummy',
	}),
};

MapDetailWidget.propTypes = {
	show: PropTypes.bool.isRequired,
	pagePosX: PropTypes.number.isRequired,
	pagePosY: PropTypes.number.isRequired,
	location: ImmutablePropTypes.contains({
		id: PropTypes.number,
		name: PropTypes.string,
	}).isRequired,
};

export default MapDetailWidget;

