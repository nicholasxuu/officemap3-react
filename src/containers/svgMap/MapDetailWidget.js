import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import {} from '../../actions/map';
import MapDetailWidget from '../../components/svgMap/MapDetailWidget';
import { getTransformMatrix, matrixMultiply, svgPosToPagePos } from '../../utils/svgUtils';

const mapStateToProps = (state) => {
	// check if we should show the object or not.
	let show = false;
	if (!state.widgetData.get('location').isEmpty()) {
		show = true;
	}

	// if show, check where in the page we should position the object
	let pagePos = {x: 0, y: 0};
	if (show === true) {
		// get final matrix
		const svgOffsetX = state.mapView.getIn(['svgOffset', 'x']);
		const svgOffsetY = state.mapView.getIn(['svgOffset', 'y']);
		const svgZoomScale = state.mapView.getIn(['zoomScale']);
		const imageWidth = state.imageDataList.get(0).get('width');
		const imageHeight = state.imageDataList.get(0).get('height');

		const transformMatrix = getTransformMatrix(svgOffsetX, svgOffsetY, svgZoomScale, imageWidth, imageHeight);

		// todo: viewport matrix can still be out of date, may need to read & update viewport matrix here instead.
		// i.e. user resized window, but never pan/zoom yet.
		const viewportMatrix = state.mapView.get('viewportMatrix');

		const finalMatrix = matrixMultiply(viewportMatrix, transformMatrix);

		const svgPos = {
			x: state.widgetData.getIn(['svgPos', 'x']),
			y: state.widgetData.getIn(['svgPos', 'y']),
		};
		pagePos = svgPosToPagePos(svgPos, finalMatrix);
	}

	return {
		show: show,
		pagePosX: pagePos.x,
		pagePosY: pagePos.y,
		location: state.widgetData.get('location'),
	};
};

const mapDispatchToProps = (dispatch) => {
	return {
		actions: bindActionCreators({}, dispatch),
	}
};

const MapDetailWidgetContainer = connect(
	mapStateToProps,
	mapDispatchToProps
)(MapDetailWidget);

export default MapDetailWidgetContainer;