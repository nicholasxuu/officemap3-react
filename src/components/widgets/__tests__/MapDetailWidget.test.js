/* eslint-disable function-paren-newline,prefer-const */
// Link.react-test.js
import React from 'react';
import Immutable from 'immutable';
import renderer from 'react-test-renderer';
import MapDetailWidget from '../MapDetailWidget';

test('DetailWidget shown', () => {
  const component = renderer.create(
    <MapDetailWidget
      show
      pagePosX={10}
      pagePosY={20}
      locationObj={Immutable.fromJS({
        id: 1,
        name: 'locationName',
        image: 'image/location/sfas.jpg',
        description: '<div>I <b>am</b> HTML</div>',
      })}
    />,
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('DetailWidget not shown', () => {
  const component = renderer.create(
    <MapDetailWidget
      show={false}
      pagePosX={10}
      pagePosY={20}
      locationObj={Immutable.fromJS({
        id: 1,
        name: 'locationName',
        image: 'image/location/sfas.jpg',
        description: '<div>I <b>am</b> HTML</div>',
      })}
    />,
  );

  let tree = component.toJSON();
  expect(tree).toBe(null);
});

test('Invalid DetailWidget not shown', () => {
  const component = renderer.create(
    <MapDetailWidget
      show
      pagePosX={10}
      pagePosY={20}
    />,
  );

  let tree = component.toJSON();
  expect(tree).toBe(null);
});
