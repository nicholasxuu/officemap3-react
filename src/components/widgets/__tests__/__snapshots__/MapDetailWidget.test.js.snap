// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailWidget shown 1`] = `
<div
  className="map-detail-widget popover top"
  id="map-detail-widget"
  role="tooltip"
  style={
    Object {
      "display": "block",
      "left": 10,
      "top": 20,
      "transform": "translateX(-50%) translateY(-100%)",
      "zIndex": 3,
    }
  }
>
  <div
    className="arrow"
    style={
      Object {
        "left": undefined,
        "top": undefined,
      }
    }
  />
  <h3
    className="popover-title"
  >
    locationName
  </h3>
  <div
    className="popover-content"
  >
    <div
      className="widget-container"
      style={
        Object {
          "display": "flex",
          "flexFlow": "row nowrap",
          "height": "100%",
          "overflow": "hidden",
          "width": "100%",
        }
      }
    >
      <div
        className="widget-image-container"
      >
        <img
          alt="locationName"
          className="widget-image"
          src="image/location/sfas.jpg"
          style={
            Object {
              "display": "block",
              "maxHeight": "100%",
              "objectFit": "cover",
              "overflow": "hidden",
              "width": "100px",
            }
          }
        />
      </div>
      <div
        className="widget-detail"
        style={
          Object {
            "display": "block",
            "overflow": "hidden",
            "width": "200px",
          }
        }
      >
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "<div>I <b>am</b> HTML</div>",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
